### PR TITLE
Add bulk utilities and adapt python3 compatibilities

### PR DESCRIPTION
**Pull request by avihad**
Add bulk utilities and adapt python3 compatibilities
- add bulk utilities #10 
- support python 3.5 changes
- add more exception
- test the new code
- add python 3.5 to travis

### CODE REVIEW

File: tests/test_bulk_utility.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+from twisted.internet.defer import succeed, inlineCallbacks
+from twisted.trial.unittest import TestCase
+
+from twistes.bulk_utils import BulkUtility, ActionParser
+from twistes.consts import EsBulk, EsDocProperties
+from twistes.exceptions import BulkIndexError, ConnectionTimeout
+
+SOME_INDEX = "some_index"
+SOME_DOC_TYPE = "some_doc_type"
+SOME_DOC = {"filed1": "value1", "field2": "value2"}
```
Comment (sxeran) [inline tests/test_bulk_utility.py]: *field1

File: twistes/compatability.py [hunk 2]
```diff
@@ -0,0 +1,13 @@
+import sys
+
+PY2 = sys.version_info[0] == 2
```
Comment (sxamit) [inline twistes/compatability.py]: sys.version_info.major == 2

File: tests/test_elasticsearch.py [hunk 3]
```diff
@@ -11,6 +11,8 @@
 from twistes.exceptions import NotFoundError, ConnectionTimeout
 from twistes.scroller import Scroller
 
+UTF_8_ENCODING = 'utf-8'
+
 FILED_2 = 'FILED_2'
 FILED_1 = 'FILED_1'
```
Comment (sxeran) [inline tests/test_elasticsearch.py]: *FIELD_1

File: tests/test_elasticsearch.py [hunk 4]
```diff
@@ -11,6 +11,8 @@
 from twistes.exceptions import NotFoundError, ConnectionTimeout
 from twistes.scroller import Scroller
 
+UTF_8_ENCODING = 'utf-8'
```
Comment (sxeran) [inline tests/test_elasticsearch.py]: Unnecessary const

File: tests/test_elasticsearch.py [hunk 5]
```diff
@@ -43,7 +45,7 @@ def setUp(self):
     def test_info(self):
         self.es._async_http_client.request = MagicMock(return_value=self.generate_response(ResponseCodes.OK))
         yield self.es.info()
-        expected_url = '{host}:{port}/'.format(host=SOME_HOST, port=SOME_PORT)
+        expected_url = '{host}:{port}/'.format(host=SOME_HOST, port=SOME_PORT).encode(UTF_8_ENCODING)
```
Comment (sxamit) [inline tests/test_elasticsearch.py]: method to create the expected url that encodes with UTF_8 and creates the url based on the**kwargs params

File: twistes/consts.py [hunk 6]
```diff
@@ -37,5 +37,35 @@ class EsConst(object):
...
+    CREATE = 'create'
+    DELETE = 'delete'
+    UPDATE = 'update'
+
+
+class EsDocProperties(object):
+    INDEX = '_index'
+    TYPE = '_type'
+    PARENT = '_parent'
+    PRECOLATE = '_percolate'
```
Comment (sxamit) [inline twistes/consts.py]: typo

File: twistes/consts.py [hunk 7]
```diff
@@ -37,5 +37,35 @@ class EsConst(object):
...
+class EsBulk(object):
+    OP_TYPE = '_op_type'
+    INDEX = 'index'
+    CREATE = 'create'
+    DELETE = 'delete'
+    UPDATE = 'update'
+
+
+class EsDocProperties(object):
+    INDEX = '_index'
```
Comment (sxamit) [inline twistes/consts.py:48]: there is no convention for the underscore consts  - 

> SCROLL_ID = 'scroll_id'
>     UNDERSCORE_SCROLL_ID = '_scroll_id'

vs

> class EsDocProperties(object):
>     INDEX = '_index'
> class EsBulk(object):
>    INDEX = 'index'

I suggest _XXX for "_xxx" (_INDEX = "_index" )

File: twistes/bulk_utils.py [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+from twistes.exceptions import BulkIndexError, ConnectionTimeout
+
+
+class ActionParser(object):
+    @staticmethod
+    def expand_action(data):
+        """
+        From one document or action definition passed in by the user extract the
+        action/data lines needed for elasticsearch's
+        :meth:`~elasticsearch.Elasticsearch.bulk` api.
```
Comment (sxamit) [inline twistes/bulk_utils.py:22]: add :return doc

File: twistes/bulk_utils.py [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+        """
+        # when given a string, assume user wants to index raw json
+        if isinstance(data, string_types):
+            return '{"index": {}}', data
+
+        # make sure we don't alter the action
+        data = data.copy()
+        op_type = data.pop(EsBulk.OP_TYPE, EsBulk.INDEX)
+        action = {op_type: {}}
+        for key in (EsDocProperties.INDEX, EsDocProperties.PARENT, EsDocProperties.PRECOLATE,
```
Comment (sxamit) [inline twistes/bulk_utils.py]: extract to method - get relevant fields, put the tuple in a variable with satisfying name

File: twistes/bulk_utils.py [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+    def bulk(self, actions, stats_only=False, **kwargs):
+        """
+        Helper for the :meth:`~elasticsearch.Elasticsearch.bulk` api that provides
+        a more human friendly interface - it consumes an iterator of actions and
+        sends them to elasticsearch in chunks. It returns a tuple with summary
+        information - number of successfully executed actions and either list of
+        errors or number of errors if `stats_only` is set to `True`.
+        See :func:`~elasticsearch.helpers.streaming_bulk` for more accepted
+        parameters
+        :arg client: instance of :class:`~elasticsearch.Elasticsearch` to use
```
Comment (sxamit) [inline twistes/bulk_utils.py]: no client is passed here (also see other methods)

File: twistes/bulk_utils.py [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+            if raise_on_exception:
+                raise e
+
+            self._handle_transport_error(bulk_actions, e, raise_on_error)
+            returnValue([])
+
+        # go through request-response pairs and detect failures
+        errors = []
+        results = []
+        for op_type, item in map(methodcaller('popitem'), resp['items']):
```
Comment (sxamit) [inline twistes/bulk_utils.py:167]: resp = None before the try

**Pull request merged**